### PR TITLE
[api] Annotate as public / move ray-core APIs to _private and add enforcement rule

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?
 Enable checking of the `ray` core module, excluding serve, workflows, and tune, in `./ci/lint/check_api_annotations.py`. This required moving many files to `ray._private` and associated fixes.

### CODE REVIEW

**Review by richardliaw — APPROVED** [review]
stamp didnt read

File: ci/lint/check_api_annotations.py [hunk 1]
```diff
@@ -13,36 +13,43 @@
...
+    try:
+        if hasattr(attr, "__module__"):
+            fullname += attr.__module__
+        if hasattr(attr, "__name__"):
+            if fullname:
+                fullname += "."
+            fullname += attr.__name__
+        if not fullname:
+            fullname = str(attr)
+    except Exception:
```
Comment (pcmoritz) [inline ci/lint/check_api_annotations.py:33]: worth logging?

File: ci/lint/check_api_annotations.py [hunk 2]
```diff
@@ -89,10 +96,12 @@ def verify(symbol, scanned, ok, output, prefix=None):
     verify(ray.rllib, set(), ok, output)
     verify(ray.air, set(), ok, output)
     verify(ray.train, set(), ok, output)
+    verify(ray, set(), ok, output, ignore=["ray.workflow", "ray.tune", "ray.serve"])
+    assert len(ok) >= 400, len(ok)
     # TODO(ekl) enable it for all modules.
     #    verify(ray.serve, set(), ok, output)
     #    verify(ray.tune, set(), ok, output)
-    #    verify(ray, set(), ok, output)
+    #    verify(ray.tune, set(), ok, output)
```
Comment (pcmoritz) [inline ci/lint/check_api_annotations.py:104]: ray.workflow?

File: python/ray/cloudpickle/cloudpickle_fast.py [hunk 3]
```diff
@@ -24,24 +24,44 @@
 
 from enum import Enum
```
Comment (pcmoritz) [inline python/ray/cloudpickle/cloudpickle_fast.py:25]: You probably want to leave this file untouched since it is copied from upstream, right?

**Review by pcmoritz — APPROVED** [review]
The cloudpickle changes should probably be reverted before merging.

**Comment by ericl** [comment]
Reverted the cloudpickle changes in favor of excluding it from the checker script for now.

**Comment by pcmoritz** [comment]
Thanks sounds great! Looks good to me, we can either try to get the codeowners or you ping me when the tests pass and we force-merge it.

**Review by clarkzinzow — APPROVED** [review]
Approving for data changes.

**Review by wuisawesome — APPROVED** [review]

**Comment by wuisawesome** [comment]
@DmitriGekhtman we should probably consider promoting `autoscaler.sdk.request_resources` to stable after this gets merged.

**Comment by DmitriGekhtman** [comment]
https://github.com/ray-project/ray/issues/22815

**Review by rkooo567 — APPROVED** [review]
We should also run nightly tests btw

**Review by gjoliver — APPROVED** [review]
RLlib is using ray._private.worker._mode() to detect whether we are in local or worker mode.
will there be a replacement API if this is explicitly marked as _private?

**Review by simon-mo — APPROVED** [review]
stamp

**Comment by ericl** [comment]
> RLlib is using ray._private.worker._mode() to detect whether we are in local or worker mode.

Given we're deprecating local mode, I think this may be something we can remove in the future.

**Pull request merged**